agent/lib/src/external_api/service.rs

```diff
@@ -0,0 +1,264 @@
+            serial_port: serialport::new(serial_path, serial_baud_rate)
+                .open()
+                .unwrap(),
+        }
+    }
+
+    pub fn read_from_serial(&mut self) -> Result<RequestMessage> {
+        info!("Reading from serial port: {}", self.serial_path);
+
+        // Create a buffer to hold the data
+        let mut data_size: usize = 0;
+        let mut size_buffer: [u8; MESSAGE_SIZE_NB_BYTES] = [0; MESSAGE_SIZE_NB_BYTES];
+        let mut got_size = false;
+
+        let mut buf = [0; 128];
+        let mut bytes_read: usize = 0;
+
+        // Create the final vector to hold the data
+        let mut data_received: Vec<u8> = Vec::new();
+
+        //we read the buffer and retrieve the first 8 bytes which are the size of the message
+        while !got_size {
+            match self.serial_port.read(&mut size_buffer) {
+                Ok(t) => {
+                    if t == 0 {
+                        break;
+                    }
+
+                    bytes_read += t;
+                    data_received.extend_from_slice(&size_buffer[..t]);
+
+                    trace!("Received {} bytes", t);
+                    trace!("Size buffer: {:?}", data_received.clone());
+
+                    if bytes_read >= MESSAGE_SIZE_NB_BYTES {
+                        got_size = true;
+
+                        let size_string = String::from_utf8(data_received.clone())
+                            .map_err(|e| anyhow!("Failed to get message size as string: {}", e))?;
+
+                        trace!("Size string: {}", size_string);
+
+                        data_size = size_string
+                            .parse::<usize>()
+                            .map_err(|e| anyhow!("Failed to parse length of message: {}", e))?;
+
+                        data_received.drain(..MESSAGE_SIZE_NB_BYTES);
+                    }
+                }
+                Err(ref e) if e.kind() == std::io::ErrorKind::TimedOut => (),
+                Err(e) => error!("{:?}", e),
+            }
+        }
+
+        bytes_read = 0;
+
+        while bytes_read < data_size {
+            match self.serial_port.read(&mut buf) {
+                Ok(t) => {
+                    if t > 0 {
+                        bytes_read += t;
+                        data_received.extend_from_slice(&buf[..t]);
+                        debug!("Received {} bytes", t);
+                    }
+                }
+                Err(ref e) if e.kind() == std::io::ErrorKind::TimedOut => (),
+                Err(e) => error!("{:?}", e),
+            }
+        }
+
+        debug!("Final received data: {:?}", data_received);
+        debug!("Total bytes read {:?}", bytes_read);
+
+        let code_entry = self.parse_json_payload(&data_received)?;
+
+        info!("Code entry: {:?}", code_entry);
+
+        // Flush the serial port
+        self.serial_port
+            .flush()
+            .map_err(|e| anyhow!("Failed to flush serial port: {}", e))?;
```

Comment:
What do we do with the flushed bytes? Maybe there were some pending bytes from the host?

Reply:
I didn't think about that. I'm deleting this